### PR TITLE
Local encryption, decryption, sign and verify.

### PR DESCRIPTION
**Pull request by m4dc4p**
* Added encrypt, decrypt, sign and verify methods to Client.
* Added sign & verify primitives to Crypto; implemented for Java and Android.
* Moved integration tests from Android to Java.
* Added simple example for encrypt/decrypt and sign/verify to README.
* Renamed `e3dbtest` to `androidtest`

### CODE REVIEW

File: e3db-crypto-plain/src/main/java/com/tozny/e3db/crypto/KaliumCrypto.java [hunk 1]
```diff
@@ -65,4 +70,57 @@ public CipherWithNonce encryptBox(byte[] message, byte[] publicKey, byte[] priva
...
+  @Override
+  public boolean verify(Signature signature, byte[] message, byte[] publicSigningKey) {
+    checkNotNull(signature, "bytes");
+    checkNotNull(message, "message");
+    checkNotNull(publicSigningKey, "publicSigningKey");
+
+    byte[] combined = new byte[signature.bytes.length + message.length];
+    System.arraycopy(signature.bytes, 0, combined, 0, signature.bytes.length);
+    System.arraycopy(message, 0, combined, signature.bytes.length, message.length);
+    return NaCl.sodium().crypto_sign_ed25519_open(new byte[message.length], null, combined, combined.length, publicSigningKey) == 0;
```
Comment (gstro) [inline e3db-crypto-plain/src/main/java/com/tozny/e3db/crypto/KaliumCrypto.java:97]: I think you can go higher-level with at least this one:
```java
return new VerifyKey(publicSigningKey).verify(message, signature.bytes)
```

File: e3db-crypto-plain/src/main/java/com/tozny/e3db/crypto/KaliumCrypto.java [hunk 2]
```diff
@@ -65,4 +70,57 @@ public CipherWithNonce encryptBox(byte[] message, byte[] publicKey, byte[] priva
...
+  @Override
+  public byte[] getPublicSigningKey(byte[] privateKey) {
+    checkNotNull(privateKey, "privateKey");
+
+    byte[] publicSigningKey = new byte[CRYPTO_SIGN_ED25519_PUBLICKEYBYTES];
+    // From libsodium's crypto_sign_ed25519_sk_to_pk (see
+    // https://github.com/jedisct1/libsodium/blob/1a3b474f7f6e6c01c56f2b3f53c1d30ed74e54ef/src/libsodium/crypto_sign/ed25519/sign_ed25519.c)
+
+    System.arraycopy(privateKey, ED25519_SEEDBYTES, publicSigningKey, 0, publicSigningKey.length);
+    return publicSigningKey;
```
Comment (gstro) [inline e3db-crypto-plain/src/main/java/com/tozny/e3db/crypto/KaliumCrypto.java:124]: That's really annoying but I'm glad you worked out the lower-level details.

**Review by gstro — APPROVED** [review]
Looks good. Nice work!

Not clear about your key management between `Client` and `Config` but I think you said you were working on it? Otherwise, mostly just superficial suggestions. 

Really disappointed in Kalium's `SigningKey` limitations.

File: e3db-crypto-android/src/main/java/com/tozny/e3db/crypto/AndroidCrypto.java [hunk 3]
```diff
@@ -70,20 +70,71 @@ public CipherWithNonce encryptBox(byte[] message, byte[] publicKey, byte[] priva
...
+    return sk;
+  }
+
+  @Override
+  public byte[] getPublicSigningKey(byte[] privateKey) {
+    byte[] dst = new byte[crypto_sign_publickeybytes()];
+    int result = crypto_sign_ed25519_sk_to_pk(dst, privateKey);
+
+    if(result != 0)
+      throw new RuntimeException("crypto_sign_ed25519_sk_to_pk: " + result);
```
[the remark this answers is not in the document]
Comment (m4dc4p) [reply e3db-crypto-android/src/main/java/com/tozny/e3db/crypto/AndroidCrypto.java:136]: Yes. Should only happen when we have a bug or something has gone terribly wrong.

File: e3db-crypto-plain/src/main/java/com/tozny/e3db/crypto/KaliumCrypto.java (hunk 1 above)
Comment (m4dc4p) [reply e3db-crypto-plain/src/main/java/com/tozny/e3db/crypto/KaliumCrypto.java:97]: That throws a RuntimeException when verification fails. Can use `try`/`catch` of course, but doesn't seem worth it.

File: e3db/src/main/java/com/tozny/e3db/Config.java [hunk 4]
```diff
@@ -115,8 +163,11 @@ public String json() {
     info.put("api_secret", apiSecret);
     info.put("client_id", clientId.toString());
     info.put("client_email", name);
-    info.put("public_key", publicKey);
+    info.put("public_key", publicEncryptionKey);
```
[the remark this answers is not in the document]
Comment (m4dc4p) [reply e3db/src/main/java/com/tozny/e3db/Config.java]: Good point.

File: e3db-crypto-plain/src/main/java/com/tozny/e3db/crypto/KaliumCrypto.java (hunk 1 above)
Comment (gstro) [reply e3db-crypto-plain/src/main/java/com/tozny/e3db/crypto/KaliumCrypto.java:97]: Ugh, gross.

**Pull request merged**